### DISCUSSION TITLE
Really love the fact that all arguments are *args and **kwargs. Who know I had the amazing ability of telepathy!

### DISCUSSION THREAD

**Discussion by user529481** [bokeh/bokeh#13869] [General]
It's so next level genius. I didn't realize how much I loved ambiguity of wanting to know what to pass to a constructor until now. 

So many years of developers wasted writing out function signatures, when we could have just had `*arg, **kwarg`. Everyone should really learn to adopt this. You know, I propose we should just name all functions variants of `fn_1, ... fn_N`. Same goes for classes, just `cls_1, ..., cls_N`. 

It is truly difficult to fathom the mind of a person who is capable of such ingenuity.

**Comment by bryevdv** [comment]
> It's so next level genius. I didn't realize how much I loved ambiguity of wanting to know what to pass to a constructor until now.
> 
> So many years of developers wasted writing out function signatures, when we could have just had `*arg, **kwarg`. Everyone should really learn to adopt this. You know, I propose we should just name all functions variants of `fn_1, ... fn_N`. Same goes for classes, just `cls_1, ..., cls_N`.
> 
> It is truly difficult to fathom the mind of a person who is capable of such ingenuity.



@user293811 We are always receptive of constructive discussion, but your post is not that. This is an official moderation with respect to our [Code of Conduct](https://github.com/bokeh/bokeh?tab=coc-ov-file): please mind your tone. 

---

There are very good reasons for the way things are set up in Bokeh, primarily due to the unusual status as a cross-runtime library that has to automatically validate and serialize all data between multiple different languages (primarily Python and JavaScript). All of the "genericity" of the APIs in support of the automation required to make that happen robustly, automatically, and without an absolute *explosion* of hard-to-maintain hand-rolled code. 

But for your information, we have in fact gone to great lengths to make sure everything is documented thoroughly.

* ALL of the thousands of Bokeh model properties, which are all the the keyword args you seem to be mystified by, are documented in the reference guide: https://docs.bokeh.org/en/latest/docs/reference/models.html This includes default information, detailed type specification, and docstring, e.g for `DataRange1d`:  

  <img width="828" alt="Screenshot 2024-05-05 at 18 48 53" src="https://github.com/bokeh/bokeh/assets/1078448/28136669-a8fe-4125-8245-1d8a0bcfd354">

* We made the effort required to dynamically generate signatures based on the property definitions so that Model initializers do show up in ipython and jupyter notebooks as "normal" signatures:
  
   ```
   In [3]: ColumnDataSource?
   Init signature:
   ColumnDataSource(
       *,
       data={},
       default_values={},
       name=None,
       selected=<Instance: Selection()>,
       selection_policy=<Instance: UnionRenderers()>,
       syncable=True,
       tags=[],
   )
   Docstring:
   Maps names of columns to sequences or arrays.
   ```

Evidently you ran into something where things don't show up as nicely?  Guess what, we could have been all too happy to discuss that specific situation, and try to see where improvements might be possible, if you had shown up behaving like a reasonable person, described what the actual situation is (amazingly enough, we are not mind-readers either), and generally started from a basis of assuming good will. 

Lastly, on a personal note: I've spent my free time helping users be successful with Bokeh for 12+ years. You have given me no reasons to want to spend my time to help you. I don't appreciate the unfounded assumptions, even less so when they are dripping with unproductive sarcasm and innuendo. I want to be clear: when you hear about "OSS burnout" or maintainers deciding to leave OSS work after decades, it is because of entitled posts like this one.

**Comment by bryevdv** [comment]
closing and locking in favor of

*  https://github.com/bokeh/bokeh/discussions/13870
